src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,240 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency"""
+
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+def RingBufferArray(size: int):
+    return RingBuffer[T](np.empty(shape=(size,), dtype=np.float64))
+
+
+
+class RingBuffer(Generic[T]):
+    def __init__(self, container: Any):
+        """Initialize the ring buffer with the given container.
+
+        Args:
+            container: Container to store the data in.
+        """
+        self.container = container
+        self.write_index = 0
+        self.read_index = 0
+        self.size = 0
+
+    def __len__(self):
+        """Return the amount of items that this container currently holds."""
+        return self.size
+
+    def max_size(self):
+        """Return the max amount of items this container can hold"""
+        return len(self.container)
+
+    def push(self, value: T) -> int:
+        """Pushes a new value into the ring buffer, returns the index in the
+        ringbuffer.
+
+        Args:
+            value: value to push into the ring buffer.
+        """
+
+        if self.size == len(self.container):
+            # Move read position one forward, dropping the oldest written value
+            self.read_index = (self.read_index + 1) % len(self.container)
+        else:
+            self.size += 1
+
+        self.container[self.write_index] = value
+        value_index = self.write_index
+        self.write_index = (self.write_index + 1) % len(self.container)
+
+        return value_index
+
+    def pop(self) -> T:
+        """Remove the oldest value from the ring buffer and return it."""
+
+        if self.size == 0:
+            raise
+
+        val = self.container[self.read_index]
+        self.read_index = (self.read_index + 1) % len(self.container)
+        self.size -= 1
+
+        return val
+
+    def wrap(self, index) -> int:
+        return index % len(self.container)
+
+    def full(self) -> bool:
+        return len(self) == len(self.container)
+
+    def __setitem__(self, index: int | slice, value: T | Sequence[T]) -> None:
+        self.container[index] = value
```

Comment:
Well, this and getitem provide a raw access to the buffer. I can certainly imagine cases where that is needed.
I can also just remove them, otherwise they aren't really needed...

Reply:
If you want to provide raw access to the underlying container, than I would make a property to get `self._container` and let people operate on that, this way is much clearer that there is no funny index adjustment going on when you slice and such. If I do:

```py
ring_buffer[0] = 1
```

I definitely expect to write the *first* element in the **ring buffer**, not the underlying container.